qp_klp/AmpliconStep.py

```diff
@@ -0,0 +1,102 @@
+from os import listdir, makedirs
+from os.path import join, isfile, basename
+from shutil import copyfile
+from qp_klp.Step import Step
+
+
+class AmpliconStep(Step):
+    def __init__(self, pipeline, master_qiita_job_id, sn_tid_map_by_project,
+                 status_update_callback=None):
+        super().__init__(pipeline,
+                         master_qiita_job_id,
+                         sn_tid_map_by_project,
+                         status_update_callback)
+
+        if pipeline.pipeline_type != 'amplicon':
+            raise ValueError("Cannot instantiate AmpliconStep object from "
+                             f"pipeline of type '{pipeline.pipeline_type}'")
+
+    def convert_bcl_to_fastq(self):
+        config = self.pipeline.configuration['bcl2fastq']
+        # note that pipeline.sample_sheet in this case
+        # is the dummy sample-sheet created by Pipeline.
+        super()._convert_bcl_to_fastq(config, self.pipeline.sample_sheet)
+
+    def quality_control(self):
```

Comment:
Something that concerns me about this is that this looks like it's completely different from other pipelines, is that right? I thought the idea was to merge it. Note that I haven't read any code below.

Reply:
The Jobs classes encapsulate all of the functionality we saw in the original shell scripts along with things we expected we'd want to have at that time. Since the actual creation and use of the Jobs classes was in qp-klp, it became the natural place to do work needed for each step and stick in additional features. A good example of this is generating failed sample records. We added it to the metagenomic pipeline but we removed it for the amplicon pipeline since we don't demultiplex the sample-names until the data is passed to Qiita.

Hence, each version of the pipeline has something they need to do at each stage; many of the stages share many of the same steps, but some need to do things before or after. In a few cases such as the QC stage of the amplicon pipeline, we need to skip adapter trimming and human filtering entirely and instead move and rename raw fastq files so they're in the form and location expected by the FastQC step.

AmpliconStep, MetagenomicStep, and future Step classes each encapsulate all the work that needs to be done at each stage. The elements common to each version are moved into the base class and called by the child classes instead of having the same code duplicated in each version.

In some cases where there's no difference between the two pipelines, they'll both just call the base class version of the step. In many cases, there'll be work that needs to be done or after the base call like generating failed sample reports in the metagenomics pipeline or setting up the ConvertJob object to use bcl2fastq instead of bcl-convert and creating a dummy sample-sheet for amplicon.

That's perhaps why the files look similar, even though there isn't much duplication of actual lines of code. The reason why the base class versions of the functions begin with '_' e.g. _quality_control() is because often the child versions have a different set of parameters and Python requires overriding methods to have the same parameter list. I could get around that with kwarg but I personally like the clarity of position parameters.

Reply:
Just to be clear, the point of this PR/work is to make this repo more python like (including nicer tests and remove duplication of code) vs. keeping the structure of the original shell script. 

Anyway, I understand your line of thought here: an object for each "Step" and have shared code in the base clase; which I guess is fine. Thank you for the explanation.